### PR TITLE
feat(unprivileged): run unprivileged nginx to support arbitrarily assigned user ID

### LINKED ISSUE #2931: Docker image doesn't/can't run as non-root user
### Is your feature request related to a problem?

We want to use `runAsNonRoot: true` in our Kubernetes cluster to prevent containers from running with root privileges. The Swagger Editor Docker image cannot run under such a configuration:

```console
$ kubectl get pods
NAME                READY   STATUS                       RESTARTS   AGE
my-swagger-editor   0/1     CreateContainerConfigError   0          2m5s

$ kubectl describe pod my-swagger-editor
Name:         my-swagger-editor
[…]
Status:       Pending
IP:           10.244.7.108
IPs:
  IP:           10.244.7.108
[…]
Containers:
  swagger-editor:
    Container ID:   
    Image:          swaggerapi/swagger-editor:v4.1.2
    […]
    State:          Waiting
      Reason:       CreateContainerConfigError
    Ready:          False
    Restart Count:  0
    […]
[…]
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  […]
  Normal   Pulled     2m14s                kubelet            Successfully pulled image "swaggerapi/swagger-editor:v4.1.2" in 3.819744701s
  Warning  Failed     2s (x12 over 2m14s)  kubelet            Error: container has runAsNonRoot and image will run as root (pod: "my-swagger-editor_swag(a5e3339c-[…])", container: swagger-editor)
  […]
```

With `runAsUser: 101` and `runAsGroup: 101`, the container can run, but crashes:

```console
$ kubectl get pods
NAME                READY   STATUS             RESTARTS   AGE
my-swagger-editor   0/1     CrashLoopBackOff   5          4m36s

$ kubectl describe pod my-swagger-editor
Name:         my-swagger-editor
[…]
Status:       Running
IP:           10.244.7.110
IPs:
  IP:           10.244.7.110
[…]
Containers:
  swagger-editor:
    Container ID:   containerd://b402034316800d47b21d8a4098c76c0aedff2e3818e1a435ba9c2ca19fb6371d
    Image:          swaggerapi/swagger-editor:v4.1.2
    […]
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Wed, 23 Feb 2022 17:37:01 +0100
      Finished:     Wed, 23 Feb 2022 17:37:01 +0100
    Ready:          False
    Restart Count:  4
    […]
[…]
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  […]
  Normal   Created    69s (x4 over 112s)   kubelet            Created container swagger-editor
  Normal   Started    69s (x4 over 111s)   kubelet            Started container swagger-editor
  Warning  BackOff    40s (x12 over 110s)  kubelet            Back-off restarting failed container
  […]

$ kubectl logs my-swagger-editor
2022/02/23 16:37:01 [emerg] 1#1: mkdir() "/var/cache/nginx/client_temp" failed (13: Permission denied)
nginx: [emerg] mkdir() "/var/cache/nginx/client_temp" failed (13: Permission denied)
```
</details>

### Describe the solution you'd like

It would be great if the Swagger Editor image didn't run as root by default, or at least didn't _have_ to run as root.

### Describe alternatives you've considered

I believe using `nginxinc/nginx-unprivileged` instead of `nginx` might be a good way to solve this. However, that would require some further modifications to the Dockerfile.

### PR DESCRIPTION
**Pull request by marius-boden-sva**
Use unprivileged nginx to support arbitrarily assigned user ID to be able to run the image e.g. on Red Hat Openshift

### Description
The dockerfile uses the `nginxinc/nginx-unprivileged` image instead of the standard `nginx` image.
Moreover the `nginx.conf` file was extended for some configuration to locate nginx runtime data in the `/tmp` directory.

This enables the container to run as a non-root user which is an improvement to the container security and enables running the container on distributions such as OpenShift by default.

### Motivation and Context
The change is necessary as the previous version required nginx to run as root to be able to read/write so several default paths.

Fixes #2931 

### How Has This Been Tested?
Dependencies: `npm ci`
Build: `npm build`
Docker Build: `docker build -t swagger-editor-unprivileged .`
Docker Run: `docker run -p 8080:8080 -u 100000 swagger-editor-unprivileged`
Visit in browser: http://localhost:8080

### My PR contains... 

- [X] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [X] Improvements (misc. changes to existing features)
- [X] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [X] are not breaking changes.

### Documentation
- [X] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [X] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by char0n** [comment]
Hi @marius-boden-sva,

Thank you for the PR! I'll have some questions on the changes to fully understand what have been changed and why if that's OK.

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM nginx:1.21.6-alpine
+FROM nginxinc/nginx-unprivileged:1.21.6-alpine
```
Comment (char0n) [inline Dockerfile:1]: This is an official image, so we should be fine. The image difference from `nginx` is as follows:

- The default NGINX listen port is now 8080 instead of 80.
- The default NGINX user directive in /etc/nginx/nginx.conf has been removed.
- The default NGINX PID has been moved from /var/run/nginx.pid to /tmp/nginx.pid. *Change _temp_path variables to /tmp/.

File: Dockerfile [hunk 2]
```diff
@@ -11,11 +11,6 @@ COPY ./dist/oauth2-redirect.html /usr/share/nginx/html/
 COPY ./dist/* /usr/share/nginx/html/dist/
 COPY ./docker-run.sh /usr/share/nginx/
 
-
-RUN chmod +x /usr/share/nginx/docker-run.sh && \
```
Comment (char0n) [inline Dockerfile:15]: Can you maybe elaborate why we're removing this chmod?

File: nginx.conf [hunk 3]
```diff
@@ -1,10 +1,21 @@
-worker_processes      1;
+worker_processes      auto;
+
+error_log  /var/log/nginx/error.log notice;
```
Comment (char0n) [inline nginx.conf:3]: Can you collaborate why changing the `error_log`?

File: nginx.conf [hunk 4]
```diff
@@ -1,10 +1,21 @@
...
+pid        /tmp/nginx.pid;
 
 events {
   worker_connections  1024;
 }
 
 http {
+  proxy_temp_path /tmp/proxy_temp;
```
Comment (char0n) [inline nginx.conf:11]: I assume these changes are all necessary to be able to run as a non root, unprivileged user, right?

File: nginx.conf [hunk 5]
```diff
@@ -1,10 +1,21 @@
...
+  client_body_temp_path /tmp/client_temp;
+  fastcgi_temp_path /tmp/fastcgi_temp;
+  uwsgi_temp_path /tmp/uwsgi_temp;
+  scgi_temp_path /tmp/scgi_temp;
+
+  include /etc/nginx/conf.d/*.conf;
```
Comment (char0n) [inline nginx.conf:17]: Why we're including the default config here?

File: nginx.conf (hunk 3 above)
Comment (marius-boden-sva) [reply nginx.conf:3]: this ist just a merge of your your configuration an the one from the `nginxinc/nginx-unprivileged:1.21.6-alpine` image

File: nginx.conf (hunk 5 above)
Comment (marius-boden-sva) [reply nginx.conf:17]: see my answerr above

File: Dockerfile (hunk 1 above)
Comment (marius-boden-sva) [reply Dockerfile:1]: The default port has to be changed because ports below 1024 are [priovileged ports](https://www.w3.org/Daemon/User/Installation/PrivilegedPorts.html#:~:text=Priviliged%20ports,has%20put%20up%20for%20you.) and require root privileges. For running the editor with docker only this results in changing the mapped port only (`-p 80:80` will become `-p 80:8080`). Reminder: need to update the docs about that ;)
For end users running swagger editor e.g. with compose or in kubernetes this will be a breaking change as they need to adapt their ports.

File: Dockerfile (hunk 2 above)
Comment (marius-boden-sva) [reply Dockerfile:15]: we do not need this as the `docker-run.sh` script is already checked into git with `-rwxr-xr-x` permissions

**Comment by marius-boden-sva** [comment]
@char0n done: https://github.com/swagger-api/swagger-editor/pull/3697

**Comment by char0n** [comment]
Superseded by https://github.com/swagger-api/swagger-editor/pull/3705
